ruff_lsp/server.py

```diff
@@ -186,38 +358,137 @@ class VersionModified(NamedTuple):
     """LSP handler for textDocument/didClose request."""
-    document = LSP_SERVER.workspace.get_text_document(params.text_document.uri)
+    text_document = LSP_SERVER.workspace.get_text_document(params.text_document.uri)
     # Publishing empty diagnostics to clear the entries for this file.
-    LSP_SERVER.publish_diagnostics(document.uri, [])
+    LSP_SERVER.publish_diagnostics(text_document.uri, [])
 
 
 @LSP_SERVER.feature(TEXT_DOCUMENT_DID_SAVE)
 async def did_save(params: DidSaveTextDocumentParams) -> None:
     """LSP handler for textDocument/didSave request."""
-    document = LSP_SERVER.workspace.get_text_document(params.text_document.uri)
-    if lint_run(_get_settings_by_document(document.path)) in ("onType", "onSave"):
-        diagnostics: list[Diagnostic] = await _lint_document_impl(document)
+    text_document = LSP_SERVER.workspace.get_text_document(params.text_document.uri)
+    if lint_run(_get_settings_by_document(text_document.path)) in (
+        Run.OnType,
+        Run.OnSave,
+    ):
+        document = Document.from_text_document(text_document)
+        diagnostics = await _lint_document_impl(document)
         LSP_SERVER.publish_diagnostics(document.uri, diagnostics)
 
 
 @LSP_SERVER.feature(TEXT_DOCUMENT_DID_CHANGE)
 async def did_change(params: DidChangeTextDocumentParams) -> None:
     """LSP handler for textDocument/didChange request."""
-    document = LSP_SERVER.workspace.get_text_document(params.text_document.uri)
-    if lint_run(_get_settings_by_document(document.path)) == "onType":
-        diagnostics: list[Diagnostic] = await _lint_document_impl(document)
+    text_document = LSP_SERVER.workspace.get_text_document(params.text_document.uri)
+    if lint_run(_get_settings_by_document(text_document.path)) == Run.OnType:
+        document = Document.from_text_document(text_document)
+        diagnostics = await _lint_document_impl(document)
         LSP_SERVER.publish_diagnostics(document.uri, diagnostics)
 
 
-async def _lint_document_impl(document: workspace.TextDocument) -> list[Diagnostic]:
+@LSP_SERVER.feature(NOTEBOOK_DOCUMENT_DID_OPEN)
+async def did_open_notebook(params: DidOpenNotebookDocumentParams) -> None:
+    """LSP handler for notebookDocument/didOpen request."""
+    notebook_document = LSP_SERVER.workspace.get_notebook_document(
+        notebook_uri=params.notebook_document.uri
+    )
+    if notebook_document is None:
+        log_warning(f"No notebook document found for {params.notebook_document.uri!r}")
+        return
+
+    document = Document.from_notebook_document(notebook_document)
+    diagnostics = await _lint_document_impl(document)
+
+    # Publish diagnostics for each cell.
+    for cell_idx, diagnostics in _group_diagnostics_by_cell(diagnostics).items():
+        LSP_SERVER.publish_diagnostics(
+            # The cell indices are 1-based in Ruff.
+            params.notebook_document.cells[cell_idx - 1].document,
+            diagnostics,
+        )
```

Comment:
The publish diagnostics capability works on a text document which, for a Notebook, is a single cell. We'll group the diagnostics by cell to make a single request for each cell.